lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1324,10 +1325,244 @@ class CopyForwardingPass : public SILFunctionTransform
+  // All normal uses (loads) must be in the initialization block.
+  // (The destroy and dealloc are commonly in a different block though.)
+  if (user->getParent() != address->getParent())
+    return false;
+
+  // Only allow uses that cannot destroy their operand. We need to be sure
+  // that replacing all this temporary's uses with the copy source doesn't
+  // destroy the source. This way, we know that the destroy_addr instructions
+  // that we recorded cover all the temporary's lifetime termination points.
+  //
+  // Currently we whitelist address projections and loads.
+  //
+  // TODO: handle non-destructive projections of enums
+  // (unchecked_take_enum_data_addr of Optional is nondestructive.)
+  switch (user->getKind()) {
+    default:
+      DEBUG(llvm::dbgs() << "  Temp use may write/destroy its source" << *user);
+      return false;
+
+    case ValueKind::StructElementAddrInst:
+    case ValueKind::TupleElementAddrInst:
+      // Transitively look through projections on stack addresses.
+      for (auto *useOper : user->getUses()) {
+        if (!collectLoads(useOper->getUser(), user, loadInsts))
+          return false;
+      }
+      return true;
+
+    case ValueKind::LoadInst:
+    case ValueKind::LoadBorrowInst:
+      // Loads are the end of the data flow chain. The users of the load can't
+      // access the temporary storage.
+      loadInsts.insert(user);
+      return true;
+
+    case ValueKind::CopyAddrInst: {
+      // copy_addr which read from the temporary are like loads.
+      // TODO: Handle copy_addr [take]. But this doesn't seem to be important.
+      auto *copyFromTmp = cast<CopyAddrInst>(user);
+      if (copyFromTmp->getDest() == address || copyFromTmp->isTakeOfSrc()) {
+        DEBUG(llvm::dbgs() << "  Temp written or taken" << *user);
+        return false;
+      }
+      loadInsts.insert(user);
+      return true;
+    }
+  }
+}
+
+/// Checks if the copy's source can be modified within the temporary's lifetime.
+///
+/// Unfortunately, we cannot simply use the destroy points as the lifetime end,
+/// because they can be in a different basic block (that's what SILGen
+/// generates). Instead we guarantee that all normal uses are within the block
+/// of the temporary and look for the last use, which effectively ends the
+/// lifetime.
+bool TempRValueOptPass::checkNoSourceModification(CopyAddrInst *copyInst,
+                     const llvm::SmallPtrSetImpl<SILInstruction *> &useInsts) {
+  unsigned NumLoadsFound = 0;
+  auto iter = std::next(copyInst->getIterator());
+  // We already checked that the useful lifetime of the temporary ends in
+  // the initialization block.
+  auto iterEnd = copyInst->getParent()->end();
+  for (; iter != iterEnd; ++iter) {
+    SILInstruction *I = &*iter;
+
+    if (useInsts.count(I))
+      NumLoadsFound++;
+
+    // If this is the last use of the temp we are ok. After this point,
+    // modifications to the source don't matter anymore.
```

Comment:
Regarding the discussion on these comments. It isn't really that simple. The order that we check for last uses and memory writes is subtle and important. It needs to be ordered this way to handle copy_addr uses, which could possibly write-back `copyDest` into `copySrc`. This won't work though if we ever handle an `apply` that takes `copyDest` as a read-only argument but may write to `copySrc`.